lang/java/reef-io/src/main/java/org/apache/reef/io/data/loading/impl/AvroEvaluatorRequestSerializer.java

```diff
@@ -0,0 +1,101 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.io.data.loading.impl;
+
+import org.apache.avro.io.*;
+import org.apache.avro.specific.SpecificDatumReader;
+import org.apache.avro.specific.SpecificDatumWriter;
+import org.apache.reef.driver.evaluator.EvaluatorRequest;
+import org.apache.reef.io.data.loading.avro.AvroEvaluatorRequest;
+import org.apache.reef.webserver.AvroHttpSerializer;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Serialize and deserialize EvaluatorRequest objects using Avro.
+ */
+public final class AvroEvaluatorRequestSerializer {
+  private static AvroEvaluatorRequest toAvro(final EvaluatorRequest request) {
+    final List<CharSequence> nodeNames = new ArrayList<>();
+    for (final String nodeName : request.getNodeNames()) {
+      nodeNames.add(nodeName);
+    }
+
+    final List<CharSequence> rackNames = new ArrayList<>();
+    for (final String rackName : request.getRackNames()) {
+      rackNames.add(rackName);
+    }
+
+    return AvroEvaluatorRequest.newBuilder()
+        .setCores(request.getNumberOfCores())
+        .setMegaBytes(request.getMegaBytes())
+        .setNumber(request.getNumber())
+        .setNodeNames(nodeNames)
+        .setRackNames(rackNames)
+        .build();
+  }
+
+  private static EvaluatorRequest fromAvro(final AvroEvaluatorRequest avroRequest) {
+    final EvaluatorRequest.Builder builder = EvaluatorRequest.newBuilder()
+        .setNumberOfCores(avroRequest.getCores())
+        .setMemory(avroRequest.getMegaBytes())
+        .setNumber(avroRequest.getNumber());
+    for (final CharSequence nodeName : avroRequest.getNodeNames()) {
+      builder.addNodeName(nodeName.toString());
+    }
+    for (final CharSequence rackName : avroRequest.getRackNames()) {
+      builder.addRackName(rackName.toString());
+    }
+    return builder.build();
+  }
+
+  public static String serialize(final EvaluatorRequest request) {
+    AvroEvaluatorRequest avroRequest = toAvro(request);
+    final DatumWriter<AvroEvaluatorRequest> datumWriter = new SpecificDatumWriter<>(AvroEvaluatorRequest.class);
+    try (final ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+      final JsonEncoder encoder = EncoderFactory.get().jsonEncoder(avroRequest.getSchema(), out);
+      datumWriter.write(avroRequest, encoder);
+      encoder.flush();
+      out.close();
+      return out.toString(AvroHttpSerializer.JSON_CHARSET);
+    } catch (final IOException ex) {
+      throw new RuntimeException("Unable to serialize compute request", ex);
+    }
+  }
+
+  public static EvaluatorRequest deserialize(final String serializedRequest) {
```

Comment:
Maybe we should have a base interface `Serializer<T>` to harmonize all of the serializers we have?